src/System.Private.CoreLib/shared/System/Text/ASCIIUtility.cs

```diff
@@ -2,75 +2,1853 @@
+
+            // JIT turns the below into constants
+
+            uint SizeOfVector128 = (uint)Unsafe.SizeOf<Vector128<byte>>();
+            nuint MaskOfAllBitsInVector128 = (nuint)(SizeOfVector128 - 1);
+
+            // This method is written such that control generally flows top-to-bottom, avoiding
+            // jumps as much as possible in the optimistic case of "all ASCII". If we see non-ASCII
+            // data, we jump out of the hot paths to targets at the end of the method.
+
+            Debug.Assert(Ssse3.IsSupported);
+            Debug.Assert(BitConverter.IsLittleEndian);
+            Debug.Assert(elementCount >= 2 * SizeOfVector128);
+
+            Vector128<short> asciiMaskForPTEST = Vector128.Create(unchecked((short)0xFF80)); // used for PTEST on supported hardware
+            Vector128<short> asciiMaskForPXOR = Vector128.Create(unchecked((short)0x8000)); // used for PXOR
+            Vector128<short> asciiMaskForPCMPGTW = Vector128.Create(unchecked((short)0x807F)); // used for PCMPGTW
+
+            // First, perform an unaligned read of the first part of the input buffer.
+
+            Vector128<short> utf16VectorHigh = Sse2.LoadVector128((short*)pUtf16Buffer); // unaligned load
+
+            // If there's non-ASCII data in the first 8 elements of the vector, there's nothing we can do.
+            // See comments in GetIndexOfFirstNonAsciiChar_Sse2 for information about how this works.
+
+            if (Sse41.IsSupported)
+            {
+                if (!Sse41.TestZ(utf16VectorHigh, asciiMaskForPTEST))
+                {
+                    return 0;
+                }
+            }
+            else
+            {
+                if (Sse2.MoveMask(Sse2.CompareGreaterThan(Sse2.Xor(utf16VectorHigh, asciiMaskForPXOR), asciiMaskForPCMPGTW).AsByte()) != 0)
+                {
+                    return 0;
+                }
+            }
+
+            // Holds the shuffle masks used to combine the first and second groups of 8 bytes into the 128-bit vector.
+
+            Vector128<byte> firstHalfMask, secondHalfMask;
+
+            // Below logic optimizes implementation of Vector128.Create so that we can
+            // get to the main logic of the method as quickly as possible.
+
+            if (IntPtr.Size >= 8)
+            {
+                firstHalfMask = Vector128.Create(0x0E0C0A08_06040200ul, 0x80808080_80808080ul).AsByte();
+                secondHalfMask = Vector128.Create(0x80808080_80808080ul, 0x0E0C0A08_06040200ul).AsByte();
+            }
+            else
+            {
+                firstHalfMask = Vector128.Create(0x00, 0x02, 0x04, 0x06, 0x08, 0x0A, 0x0C, 0x0E, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80);
+                secondHalfMask = Vector128.Create(0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x00, 0x02, 0x04, 0x06, 0x08, 0x0A, 0x0C, 0x0E);
+            }
+
+            // Turn the 8 ASCII cha
```

Comment:
Is there big-endian Ssse3? 🤔

Reply:
Time to round up some stray if checks. :)